File: Laboratorio7/Herencia.java
package laboratorio7;

class Empleado {
    private String nombre;
    private String ID;
    private double salarioBase;

    public Empleado(String nombre, String ID, double salarioBase) {
        this.nombre = nombre;
        this.ID = ID;
        this.salarioBase = salarioBase;
    }

    public void setNombre(String nombre) {
        this.nombre = nombre;
    }
    
    public void setID(String ID) {
        this.ID = ID;
    }

    public void setSalarioBase(double salarioBase) {
        this.salarioBase = salarioBase;
    }

    public String getNombre() {
        return nombre;
    }

    public String getID() {
        return ID;
    }

    public double calcularSalario() {
        return salarioBase;
    }

    public String getInfo() {
        return String.format("El empleado: %s \nCon ID: %s \nTiene el salario: %.2f", getNombre(), getID(), calcularSalario());
    }
}

class EmpleadoTiempoCompleto extends Empleado {
    public EmpleadoTiempoCompleto(Empleado empleado) {
        super(empleado.getNombre(), empleado.getID(), empleado.calcularSalario());
    }

    @Override
    public double calcularSalario() {
        return super.calcularSalario() * 1.1;
    }

    @Override
    public String getInfo() {
        return String.format("El empleado: %s \nCon ID: %s \nSalario calculado (Tiempo Completo): %.2f\n", getNombre(), getID(), calcularSalario());
    }
}

class EmpleadoMedioTiempo extends Empleado {
    public EmpleadoMedioTiempo(Empleado empleado) {
        super(empleado.getNombre(), empleado.getID(), empleado.calcularSalario());
    }

    @Override
    public double calcularSalario() {
        return super.calcularSalario() * 0.5;
    }

    @Override
    public String getInfo() {
        return String.format("El empleado: %s \nCon ID: %s \nSalario calculado (Medio Tiempo): %.2f\n", getNombre(), getID(), calcularSalario());
    }
}

class EmpleadoPorHoras extends Empleado {
    private int horasTrabajadas;

    public EmpleadoPorHoras(Empleado empleado, int horasTrabajadas) {
        super(empleado.getNombre(), empleado.getID(), empleado.calcularSalario());
        this.horasTrabajadas = horasTrabajadas;
    }

    public void setHorasTrabajadas(int horasTrabajadas) {
        this.horasTrabajadas = horasTrabajadas;
    }

    public int getHorasTrabajadas() {
        return horasTrabajadas;
    }

    @Override
    public double calcularSalario() {
        return super.calcularSalario() * getHorasTrabajadas();
    }

    @Override
    public String getInfo() {
        return String.format("El empleado: %s \nCon ID: %s \nSalario calculado (Por Horas): %.2f\n", getNombre(), getID(), calcularSalario());
    }
}

public class Herencia {
    public static void main(String[] args) {
        Empleado empleado1 = new Empleado("Manuel", "34F24N51", 1500.0);
        EmpleadoTiempoCompleto ETC = new EmpleadoTiempoCompleto(empleado1);

        Empleado empleado2 = new Empleado("Jose", "1F4N6B8A", 1500.0);
        EmpleadoMedioTiempo EMT = new EmpleadoMedioTiempo(empleado2);

        Empleado empleado3 = new Empleado("Marcelo", "S1M3K4L1", 200.0);
        EmpleadoPorHoras EPH = new EmpleadoPorHoras(empleado3, 8);

        Empleado[] listaEmpleados = {ETC, EMT, EPH};

        for (Empleado empleado : listaEmpleados) {
            System.out.println(empleado.getInfo());
        }
    }
}

File: Laboratorio7/Herencia.py
class Empleado:
    def __init__(self, nombre, ID, salarioBase) -> None:
        self._nombre = nombre
        self._ID = ID
        self._salarioBase = salarioBase
        pass

    def setNombre(self, n_nombre):
        self._nombre = n_nombre

    def setID(self, n_ID):
        self._ID = n_ID

    def setSalarioBase(self, n_salarioBase):
        self._salarioBase = n_salarioBase

    def getNombre(self):
        return self._nombre

    def getID(self):
        return self._ID

    def calcularSalario(self):
        return self._salarioBase
    
    def getInfo(self):
        return f"El empleado: {self.getNombre()} \nCon ID: {self.getID()} \nTiene el salario: {self.calcularSalario():.2f}"
    
class EmpleadoTiempoCompleto(Empleado):
    def __init__(self, empleado) -> None:
        super().__init__(empleado.getNombre(), empleado.getID(), empleado.calcularSalario())

    def calcularSalario(self):
        return super().calcularSalario()*1.1
    
    def getInfo(self):
        return f"El empleado: {self.getNombre()} \nCon ID: {self.getID()} \nSalario calculado (Tiempo Completo): {self.calcularSalario():.2f}\n"
    
class EmpleadoMedioTiempo(Empleado):
    def __init__(self, empleado) -> None:
        super().__init__(empleado.getNombre(), empleado.getID(), empleado.calcularSalario())

    def calcularSalario(self):
        return super().calcularSalario()*.5
    
    def getInfo(self):
        return f"El empleado: {self.getNombre()} \nCon ID: {self.getID()} \nSalario calculado (Medio Tiempo): {self.calcularSalario():.2f}\n"
    
class EmpleadoPorHoras(Empleado):
    def __init__(self, empleado, horasTrabajadas) -> None:
        super().__init__(empleado.getNombre(), empleado.getID(), empleado.calcularSalario())
        self._horasTrabajadas = horasTrabajadas

    def setHorasTrabajadas(self, n_horasTrabajadas):
        self._horasTrabajadas = n_horasTrabajadas

    def getHorasTrabajadas(self):
        return self._horasTrabajadas

    def calcularSalario(self):
        return super().calcularSalario()*self.getHorasTrabajadas()
    
    def getInfo(self):
        return f"El empleado: {self.getNombre()} \nCon ID: {self.getID()} \nSalario calculado (Por Horas): {self.calcularSalario():.2f}\n"
    

if __name__ == '__main__':
    empleado1 = Empleado("Manuel", "34F24N51", 1500.0)
    ETC = EmpleadoTiempoCompleto(empleado1)

    empleado2 = Empleado("Jose", "1F4N6B8A", 1500.0)
    EMT = EmpleadoMedioTiempo(empleado2)

    empleado3 = Empleado("Marcelo", "S1M3K4L1", 200.0)
    EPH = EmpleadoPorHoras(empleado3,8)

    lista_empleados = [ETC, EMT, EPH]

    for empleado in lista_empleados:
        print(empleado.getInfo())